pkg/provisioner/provisioner.go

```diff
@@ -0,0 +1,273 @@
+	Serve(ctx context.Context) error
+	ListPods(ctx context.Context) error
+	DeletePod(ctx context.Context, UID string) error
+}
+
+type k8sProvisioner struct {
+	Namespace  string
+	Clientset  *kubernetes.Clientset
+	Pods       corev1.PodInterface
+	restConfig *rest.Config
+	events     corev1.EventInterface
+}
+
+// NewK8sProvisioner returns a new provisioner implementation that targets
+// Kubernetes using a specific Kubernetes namespace and REST config.
+func NewK8sProvisioner(namespace string, restConfig *rest.Config) (Provisioner, error) {
+	clientset, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	return k8sProvisioner{
+		Namespace:  namespace,
+		Clientset:  clientset,
+		Pods:       clientset.CoreV1().Pods(namespace),
+		restConfig: restConfig,
+		events:     clientset.CoreV1().Events(namespace),
+	}, nil
+}
+
+func (p k8sProvisioner) ListPods(ctx context.Context) error {
+	podList, err := p.Pods.List(ctx, metav1.ListOptions{
+		LabelSelector: "app.kubernetes.io/name=wharf-cmd-worker,app.kubernetes.io/managed-by=wharf-cmd-provisioner,wharf.iver.com/instance=prod",
+	})
+	if err != nil {
+		return err
+	}
+	pods := podList.Items
+
+	printPod := func(pod metav1.ObjectMeta, i int) func(logger.Event) logger.Event {
+		return func(ev logger.Event) logger.Event {
+			ev = ev.WithInt("index", i).
+				WithString("UID", string(pod.UID)).
+				WithString("name", pod.Name).
+				WithString("namespace", pod.Namespace)
+
+			for k, v := range pod.Labels {
+				ev = ev.WithString(k, v)
+			}
+
+			return ev
+		}
+	}
+
+	log.Info().WithInt("count", len(pods)).Message("Fetched pods with matching labels.")
+	for i, pod := range pods {
+		log.Info().WithFunc(printPod(pod.ObjectMeta, i)).Message("Pod")
+	}
+
+	return nil
+}
+
+func (p k8sProvisioner) DeletePod(ctx context.Context, UID string) error {
+	podList, err := p.Pods.List(ctx, metav1.ListOptions{
+		LabelSelector: "app.kubernetes.io/name=wharf-cmd-worker,app.kubernetes.io/managed-by=wharf-cmd-provisioner,wharf.iver.com/instance=prod",
+	})
+	if err != nil {
+		return err
+	}
+
+	var matchingPod *v1.Pod
+	for _, pod := range podList.Items {
+		if string(pod.ObjectMeta.UID) == UID {
+			matchingPod = &pod
+			break
+		}
+	}
+
+	if matchingPod == nil {
+		return fmt.Errorf("found no pod with appropriate labels matching UID: %s", UID)
+	}
+
+	return p.Pods.Delete(ctx, matchingPod.Name, metav1.DeleteOptions{})
+}
+
+func (p k8sProvisioner) Serve(ctx context.Context) error {
+	podMeta := createPodMeta()
+
+	newPod, err := p.Pods.Create(ctx, &podMeta, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	err = p.waitForInitContainerDone(ctx, newPod.ObjectMeta)
+	if err != nil {
+		return err
+	}
+
+	log.Debug().Message("Init Container done.")
+
+	err = p.waitForAppContainerRunning(ctx, newPod.ObjectMeta)
+	if err != nil {
+		return err
+	}
+
+	log.Debug().Message("App Container running.")
+
+	err = p.streamLogsUntilCompleted(ctx, newPod.Name)
+	return err
+}
```

Comment:
Maybe this `Serve` should be its own package. Like `provisionerapi`. It's such a different use case to serve a HTTP REST API.

Also, this `Serve` implementation doesn't do much of serving than it is creating. So perhaps renaming this method to `CreateWorker` and let a package such as `provisionerapi` deal with the HTTP serving stuff

Reply:
Oh, yeah, I was under the impression for whatever reason that `serve` was meant to create the workers, not launch the HTTP server 😅

Side-note: Should I add a command for `create` as I have done for `delete` and `list`, or remove all the commands except `serve`? I liked being able to do it via commands, and it feels a bit more complete being able to do it without API calls.

I like the separation of api and logic into different packages so will do that. 👍🏻

Reply:
Yea I mean the commands are great. It's the wharf-api that will talk to wharf-cmd-provisioner, but having the commands doesn't hurt as it stands right now.

I'm a little bit worried about the feeding of build args and Git credentials. Different problem for a different day /shrug